### PR TITLE
【KernelGen】Add logsumexp operator

### PR DESCRIPTION
**Pull request by Schopenhauer-loves-Hegel**
### PR Category
Operator

### Type of Change
New Feature

### Description
Add `logsumexp` operator implementation with Triton kernel.

- Implementation mode: `N/A`
- Accuracy test: 36/36 passed

### Issue
N/A

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### Performance
**torch.bfloat16**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| torch.Size([64, 64]) | 0.0359 | 0.0076 | 4.730 |
| torch.Size([256, 256]) | 0.0373 | 0.0083 | 4.519 |
| torch.Size([1024, 1024]) | 0.0484 | 0.0115 | 4.223 |
| torch.Size([4096, 4096]) | 0.2187 | 0.0467 | 4.680 |
| torch.Size([1024, 65536]) | 0.7471 | 0.1287 | 5.805 |
| torch.Size([10000, 1]) | 0.0381 | 0.0153 | 2.492 |
| torch.Size([10000, 256]) | 0.0609 | 0.0194 | 3.147 |
| torch.Size([10000, 65536]) | 6.6176 | 0.9506 | 6.961 |

**torch.float16**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| torch.Size([64, 64]) | 0.0448 | 0.0089 | 5.040 |
| torch.Size([256, 256]) | 0.0370 | 0.0084 | 4.433 |
| torch.Size([1024, 1024]) | 0.0471 | 0.0109 | 4.304 |
| torch.Size([4096, 4096]) | 0.2161 | 0.0466 | 4.636 |
| torch.Size([1024, 65536]) | 0.7437 | 0.1271 | 5.851 |
| torch.Size([10000, 1]) | 0.0383 | 0.0153 | 2.502 |
| torch.Size([10000, 256]) | 0.0598 | 0.0188 | 3.179 |
| torch.Size([10000, 65536]) | 6.5999 | 0.9473 | 6.967 |

**torch.float32**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| torch.Size([64, 64]) | 0.0348 | 0.0076 | 4.582 |
| torch.Size([256, 256]) | 0.0365 | 0.0083 | 4.392 |
| torch.Size([1024, 1024]) | 0.0477 | 0.0135 | 3.527 |
| torch.Size([4096, 4096]) | 0.3537 | 0.0737 | 4.799 |
| torch.Size([1024, 65536]) | 1.2220 | 0.2185 | 5.592 |
| torch.Size([10000, 1]) | 0.0365 | 0.0154 | 2.375 |
| torch.Size([10000, 256]) | 0.0650 | 0.0204 | 3.188 |
| torch.Size([10000, 65536]) | 11.3369 | 1.8290 | 6.198 |

**Overall: median speedup = 4.550x, mean speedup = 4.505x** (24 data points)

---
_Generated by auto_gen tool with Claude Code_

### CODE REVIEW

**Comment by factnn** [comment]
## logsumexp Benchmark Speedup Results

### float16
| Size | Torch (ms) | Gems (ms) | Speedup |
|------|-----------|----------|---------|
| [64, 64], dim=1 | 0.035 | 0.008 | 4.375 |
| [256, 256], dim=1 | 0.037 | 0.008 | 4.625 |
| [1024, 1024], dim=1 | 0.047 | 0.012 | 3.917 |
| [4096, 4096], dim=1 | 0.218 | 0.047 | 4.638 |
| [1024, 65536], dim=1 | 0.746 | 0.131 | 5.695 |
| [10000, 1], dim=1 | 0.041 | 0.015 | 2.690 |
| [10000, 256], dim=1 | 0.060 | 0.019 | 3.110 |
| [10000, 65536], dim=1 | 6.593 | 0.993 | 6.642 |

### float32
| Size | Torch (ms) | Gems (ms) | Speedup |
|------|-----------|----------|---------|
| [64, 64], dim=1 | 0.035 | 0.007 | 4.769 |
| [256, 256], dim=1 | 0.036 | 0.009 | 4.011 |
| [1024, 1024], dim=1 | 0.049 | 0.013 | 3.802 |
| [4096, 4096], dim=1 | 0.355 | 0.074 | 4.794 |
| [1024, 65536], dim=1 | 1.228 | 0.220 | 5.587 |
| [10000, 1], dim=1 | 0.040 | 0.015 | 2.585 |
| [10000, 256], dim=1 | 0.067 | 0.020 | 3.342 |
| [10000, 65536], dim=1 | 11.380 | 1.835 | 6.203 |

### bfloat16
| Size | Torch (ms) | Gems (ms) | Speedup |
|------|-----------|----------|---------|
| [64, 64], dim=1 | 0.036 | 0.009 | 4.187 |
| [256, 256], dim=1 | 0.038 | 0.008 | 4.841 |
| [1024, 1024], dim=1 | 0.048 | 0.012 | 4.000 |
| [4096, 4096], dim=1 | 0.219 | 0.047 | 4.681 |
| [1024, 65536], dim=1 | 0.750 | 0.131 | 5.716 |
| [10000, 1], dim=1 | 0.040 | 0.014 | 2.805 |
| [10000, 256], dim=1 | 0.061 | 0.019 | 3.239 |
| [10000, 65536], dim=1 | 6.611 | 0.984 | 6.720 |

File: benchmark/test_logsumexp_perf.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import pytest
```
Comment (tengqm) [inline benchmark/test_logsumexp_perf.py]: remove `_perf` from the file name.

File: benchmark/test_logsumexp_perf.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import torch
+
+import flag_gems
+from benchmark.attri_util import FLOAT_DTYPES
+from benchmark.performance_utils import GenericBenchmark, generate_tensor_input
```
Comment (tengqm) [inline benchmark/test_logsumexp_perf.py]: Rebase to master, `performance_utils` and `attri_util` has been renamed.

File: src/flag_gems/ops/__init__.py [hunk 3]
```diff
@@ -741,20 +711,15 @@
...
     "upsample_nearest1d",
     "upsample_nearest2d",
     "upsample_nearest3d",
     "var_mean",
-    "var",
-    "var_correction",
-    "var_dim",
     "vdot",
     "vector_norm",
     "vstack",
```
Comment (tengqm) [inline src/flag_gems/ops/__init__.py:793]: The above changes are insane.

File: src/flag_gems/__init__.py [hunk 4]
```diff
@@ -461,16 +441,12 @@ def torch_ge(v):
...
     ("upsample_nearest1d", upsample_nearest1d),
     ("upsample_nearest2d", upsample_nearest2d),
     ("upsample_nearest3d", upsample_nearest3d),
     ("var_mean.correction", var_mean),
-    ("var", var),
-    ("var.correction", var_correction),
-    ("var.dim", var_dim),
     ("vdot", vdot),
     ("vstack", vstack),
     ("where.self", where_self),
```
Comment (tengqm) [inline src/flag_gems/__init__.py:496]: The above changes are insane.

File: tests/test_logsumexp.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+from .accuracy_utils import (
+    FLOAT_DTYPES,
+    REDUCTION_SHAPES,
+    gems_assert_close,
+    to_reference,
+)
+from .conftest import QUICK_MODE
+
+DIM_LIST = [0] if QUICK_MODE else [0, 1]
```
Comment (tengqm) [inline tests/test_logsumexp.py:14]: Looks to me that this `quick` mode is not quick enough. Can be dropped.

**Comment by factnn** [comment]
## UT Results

Branch rebased to upstream/master and tests passed:

```
python -m pytest tests/ -m logsumexp --ref cpu -x -q
36 passed
```

**Comment by factnn** [comment]
Both suggestions have been addressed: the benchmark file has been renamed to test_logsumexp.py (removing_perf), and the imports have been updated to use from . import base, consts, utils after rebasing to master.

**Review by w1120029931-bit — APPROVED** [review]

**Review by bin913 — APPROVED** [review]
